src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -191,197 +371,190 @@ private ExecutorInfo createExecutor(String taskIdName, String nodeName, String e
+          offer,
+          HDFSConstants.JOURNAL_NODE_ID,
+          Arrays.asList(HDFSConstants.JOURNAL_NODE_ID),
+          HDFSConstants.NODE_EXECUTOR_ID);
+      return true;
+    }
+    return false;
   }
 
-  @Override
-  synchronized public void resourceOffers(SchedulerDriver driver, List<Offer> offers) {
-    log.info(String.format("Received %d offers", offers.size()));
-    //TODO(elingg) all datanodes can be launched together after the other nodes have initialized.
-    //Remove this waiting period for datanodes.
-    if (!liveState.getStagingTasks().isEmpty()) {
-      log.info("Declining offers because tasks are currently staging");
-      for (Offer offer : offers) {
-        driver.declineOffer(offer.getId());
-      }
-      return;
+  private boolean tryToLaunchNameNode(SchedulerDriver driver, Offer offer) {
+    if (offerNotEnoughResources(offer,
+        (conf.getNameNodeCpus() + conf.getZkfcCpus()),
+        (conf.getNameNodeHeapSize() + conf.getZkfcHeapSize()))) {
+      log.info("Offer does not have enough resources");
+      return false;
     }
 
-    if (liveState.getNameNodes().size() == 0 && liveState.getJournalNodes().size() == 0) {
-      log.info("No NameNodes or JournalNodes found.  Collecting offers until we have sufficient "
-          + "capacity to launch.");
-      for (Offer offer: offers) {
-          pendingOffers.put(offer.getId(), offer);
-      }
-        
-      if (!initializingCluster) {
-        log.info(String.format("Launching initial nodes with %d pending offers",
-            pendingOffers.size()));
-        initializingCluster = true;
-        launchInitialJournalNodes(driver, pendingOffers.values());
-        launchInitialNameNodes(driver, pendingOffers.values());
-        // Decline any remaining offer
-        for (OfferID offerID : pendingOffers.keySet()) {
-          driver.declineOffer(offerID);
-        }
-        pendingOffers.clear();
+    boolean launch = false;
+    List<String> deadNameNodes = persistentState.getDeadNameNodes();
+
+    if (deadNameNodes.isEmpty()) {
+      if (liveState.getNameNodeSize() >= HDFSConstants.TOTAL_NAME_NODES) {
+        log.info(String.format("Already running %s namenodes", HDFSConstants.TOTAL_NAME_NODES));
+      } else if (persistentState.nameNodeRunningOnSlave(offer.getHostname())) {
+        log.info(String.format("Already running namenode on %s", offer.getHostname()));
+      } else if (persistentState.dataNodeRunningOnSlave(offer.getHostname())) {
+        log.info(String.format("Cannot colocate namenode and datanode on %s", offer.getHostname()));
+      } else if (!persistentState.journalNodeRunningOnSlave(offer.getHostname())) {
+        log.info(String.format("We need to coloate the namenode with a journalnode and there is"
+            + "no journalnode running on this host. %s", offer.getHostname()));
+      } else {
+        launch = true;
       }
-      return;
+    } else if (deadNameNodes.contains(offer.getHostname())) {
```

Comment:
Again, why restrict ourselves to one of the deadNN hosts, especially if that machine is never coming back alive? Wouldn't any current JN do?